server/src/test/java/io/crate/expression/scalar/ArraySliceFunctionTest.java

```diff
@@ -0,0 +1,165 @@
+    }
+
+    @Test
+    public void testToIsMoreThanSizeOfArray() {
+        assertEvaluate("[1, 2, 3, 4, 5][ 3 :100]", List.of(3, 4, 5));
+    }
+
+    @Test
+    public void testFromAndToIsMoreThanSizeOfArray() {
+        assertEvaluate("[1, 2, 3, 4, 5][ 20:100 ]", List.of());
+    }
+
+    @Test
+    public void testFromIsBiggerThanTo() {
+        assertEvaluate("[1, 2, 3, 4, 5][ 3 : 1 ]", List.of());
+    }
+
+    @Test
+    public void testEmptyArray() {
+        assertEvaluate("([]::array(integer))[1 :10]", List.of());
+    }
+
+    @Test
+    public void testNullArray() {
+        assertEvaluate("(null::array(integer))[1 :10]", List.of());
+    }
+
+    @Test
+    public void testFromIsNull() {
+        assertEvaluate("[1,2,3,4,5][null:3]", List.of(1, 2, 3));
+    }
+
+    @Test
+    public void testToIsNull() {
+        assertEvaluate("[1,2,3,4,5][3:null]", List.of(3, 4, 5));
+    }
+
+    @Test
+    public void testFromIsNotAnInteger() {
+        expectedException.expect(ConversionException.class);
+        expectedException.expectMessage("Cannot cast `'three'` of type `text` to type `integer`");
+        assertEvaluate("[1,2,3,4,5]['three':]", null);
+    }
+
+    @Test
+    public void testToIsNotAnInteger() {
+        expectedException.expect(ConversionException.class);
+        expectedException.expectMessage("Cannot cast `'three'` of type `text` to type `integer`");
+        assertEvaluate("[1,2,3,4,5][:'three']", null);
+    }
+
+    @Test
+    public void testBaseIsNotAnArray() {
+        expectedException.expect(UnsupportedOperationException.class);
+        expectedException.expectMessage("An expression of type StringLiteral cannot have " +
+                                        "an array slice ([<from>:<to>])");
+        assertEvaluate("'not an array'[1:3]", null);
+    }
+
+    @Test
+    public void testFromIsNegative() {
+        expectedException.expect(UnsupportedOperationException.class);
+        expectedException.expectMessage("Array index must be in range 1 to 2147483648");
+        assertEvaluate("[1,2,3,4,5][-1:]", null);
+    }
+
+    @Test
+    public void testToIsNegative() {
+        expectedException.expect(UnsupportedOperationException.class);
+        expectedException.expectMessage("Array index must be in range 1 to 2147483648");
+        assertEvaluate("[1,2,3,4,5][:-1]", null);
+    }
+
+//    @Test
+//    public void testFromIsBig() {
+//        expectedException.expect(UnsupportedOperationException.class);
+//        expectedException.expectMessage("Array index must be in range 1 to 2147483648");
+//        assertEvaluate("[1,2,3,4,5][2147483648:]", List.of());
+//    }
+//
+//    @Test
+//    public void testToIsBig() {
+//        expectedException.expect(UnsupportedOperationException.class);
+//        expectedException.expectMessage("Array index must be in range 1 to 2147483648");
+//        assertEvaluate("[1,2,3,4,5][:2147483648]", List.of(1, 2, 3, 4, 5));
+//    }
```

Comment:
These tests are commented out because they don't pass in the current implementation.
The current `ArraySliceValidator` follows the `SubscriptValidator` logic for having the max value `Integer.MAX_VALUE + 1L`. 
The problem is that ` Integer.MAX_VALUE + 1L` is already a long number, not an integer. The cast to `integer` fails https://github.com/crate/crate/blob/b8f5bfc1ae1d34b0426f2a01217fde663505225b/server/src/main/java/io/crate/analyze/SubscriptValidator.java#L146.
So currently this subscript access doesn't work `arr[2147483648]`. It fails with the error:
```
 Cannot cast `2147483648::bigint` of type `bigint` to type `integer`
```

I see a couple of solutions:
1) decrease the max value for the index in both the subscript expression and the slice to `2147483647` and update the docs accordingly.
2) cast to the `bigint`. This could require using longs before transforming indices to start from `0` https://github.com/crate/crate/blob/3169e4ac6eddccc48d7ed1a6b197fa0e1ac2152d/server/src/main/java/io/crate/expression/scalar/SubscriptFunction.java#L208. Also I am not sure what the performance implication of such change will be.

Reply:
Good catch!
Lets do 1), decrease range by one, it's not a bigger deal and avoid this hacky workaround of int<->long conversions.
I'll fix this for subscript as well.

Reply:
Right good catch again.
If the slice lower/upper bound expression evaluates to a `Literal` we could use `compile`, otherwise we really must validate on every `evaluate` and such *row*.

Another simple solution could be to forbid using expressions as slice bounds at the `ArraySliceValidator`.
I'm fine with both ways, although the later limits the functionality, but also not sure if using expressions is a common use-case.

Reply:
I implemented a double validation: 
- first in the Validator and the Long values. This prevents `ConversionException` from happening when passing the long into the the `ArraySliceFuction`
- the int is validated to be 1 or more in the `ArraySliceFunction` itself.